atoma-state/src/state_manager.rs

```diff
@@ -3692,6 +3693,161 @@ impl AtomaState {
+    ///    state_manager.get_sui_address(sui_address).await
+    /// }
+    /// ```
+    #[instrument(level = "trace", skip(self))]
+    pub async fn get_user_id(&self, sui_address: String) -> Result<Option<i64>> {
+        let user = sqlx::query("SELECT id FROM users WHERE sui_address = $1")
+            .bind(sui_address)
+            .fetch_optional(&self.db)
+            .await?;
+
+        Ok(user.map(|user| user.get("id")))
+    }
+
+    /// Update the balance for the user.
+    ///
+    /// This method updates the `balance` field for the user in the `users` table.
+    ///
+    /// # Arguments
+    ///
+    /// * `user_id` - The unique identifier of the user.
+    /// * `balance` - The new balance to store for the user.
+    ///
+    /// # Returns
+    ///
+    /// - `Result<()>`: A result indicating success (Ok(())) or failure (Err(AtomaStateManagerError)).
+    ///
+    /// # Errors
+    ///
+    /// This function will return an error if:
+    ///
+    /// - The database query fails to execute.
+    ///
+    /// # Example
+    ///
+    /// ```rust,ignore
+    /// use atoma_node::atoma_state::AtomaStateManager;
+    ///
+    /// async fn update_balance(state_manager: &AtomaStateManager, user_id: i64, balance: i64) -> Result<(), AtomaStateManagerError> {
+    ///    state_manager.update_balance(user_id, balance).await
+    /// }
+    /// ```
+    #[instrument(level = "trace", skip(self))]
+    pub async fn top_up_balance(&self, user_id: i64, balance: i64, timestamp: i64) -> Result<()> {
+        sqlx::query(
+            "INSERT INTO balance (user_id, usdc_balance, usdc_last_timestamp) 
+                         VALUES ($1, $2, $3) 
+                         ON CONFLICT (user_id) 
+                         DO UPDATE SET 
+                            usdc_balance = balance.usdc_balance + EXCLUDED.usdc_balance, 
+                            usdc_last_timestamp = EXCLUDED.usdc_last_timestamp 
+                         WHERE balance.usdc_last_timestamp < EXCLUDED.usdc_last_timestamp",
+        )
+        .bind(user_id)
+        .bind(balance)
+        .bind(timestamp)
+        .execute(&self.db)
+        .await?;
+        Ok(())
+    }
+
+    /// Withdraw balance for the user.
+    ///
+    /// This method withdraws the balance from the user in the `users` table.
+    ///
+    /// # Arguments
+    ///
+    /// * `user_id` - The unique identifier of the user.
+    /// * `balance` - The balance to withdraw from the user.
+    ///
+    /// # Returns
+    ///
+    /// - `Result<()>`: A result indicating success (Ok(())) or failure (Err(AtomaStateManagerError)).
+    ///
+    /// # Errors
+    ///
+    /// This function will return an error if:
+    ///
+    /// - The database query fails to execute (that could mean the balance is not available)
+    pub async fn withdraw_balance(&self, user_id: i64, balance: i64) -> Result<()> {
```

Comment:
I would rename the method name, `withdraw_balance` gives the impression that we really are withdrawing funds

Reply:
Renamed to deduct_from_usdc